### ISSUE TITLE
SDL 2.30.9-1 | Broken HW renderer in Mesen2

### ISSUE THREAD

**Issue by LimitSeeker** [libsdl-org/SDL#11710]
(as detailed in #8340), it will appear as if no game was loaded. I'm on Nobara (based on Fedora)

![image](https://github.com/user-attachments/assets/e97805cc-9218-43ef-a45b-d2a216aabc6a)
![image](https://github.com/user-attachments/assets/d404f92f-5a3b-42fe-bbc5-8bcd809785a2)

**Comment by slime73** [comment]
The other issue you linked supposedly ended up being a problem with Mesa rather than SDL. Have you ruled that out for yourself?

**Comment by LimitSeeker** [comment]
I've switched from Mesa stable to git build, and it still does persist, not sure tbh

**Comment by Squall-Leonhart** [comment]
Multi-gpu system and the one used by mesen isn't 'awake'?

**Comment by LimitSeeker** [comment]
No, unfortunately, dedicated GPU, the CPU doesn't have an integrated GPU either, thx nonetheless!

**Comment by LimitSeeker** [comment]
Update: If you run Mesen 2 through gamescope the video renders without toggling SW mode on; how odd...

**Comment by SourMesen** [comment]
@LimitSeeker It looks like the issue is somehow related to Wayland (but I'm unsure where the actual problem is). I tweaked some things in Mesen's code to force SDL2 to use its X11 video driver when the default driver fails, and it seems to fix things on my end on a Fedora VM. So with some luck it should work on your end if you get the latest build, without having to enable the software renderer.

**Comment by icculus** [comment]
[We discussed this briefly in sdl2-compat](https://github.com/libsdl-org/sdl2-compat/issues/262#issuecomment-2672370788)...Mesen hands SDL a GTK+ window (via Avalonia) to use with SDL_CreateWindowFrom...it's possible the GTK+ Wayland backend is handing us something other than we want (the wrong surface, some other thing?), or that Mesen is using XWayland in the first place, so this completely blows up if SDL initializes for Wayland.

CC'ing @Kontrabant so he's aware, but I don't have any further useful information at the moment.

**Comment by Kontrabant** [comment]
Yep, the app presumes X11 and defaulting to Wayland will break things. It looks like Mesen added a workaround for this a few days ago: https://github.com/SourMesen/Mesen2/commit/97d9b05c742bd5dcfb79949b52c9c09e0910b68f

**Comment by icculus** [comment]
I do feel like we need to figure out what GTK+/wayland hands us here, at some point, but for now, is this resolved to everyone's satisfaction?

**Comment by Kontrabant** [comment]
> I do feel like we need to figure out what GTK+/wayland hands us here,

SDL3 can take external surfaces from Qt/GTK/whatever for rendering, and there is a simple example of creating a window in Qt and using SDL for rendering in README-wayland.md.

As long as the toolkit lets you get the wl_display and wl_surface, SDL3 can interoperate with it.

**Comment by SourMesen** [comment]
I'm not a Linux user in general, so I probably can't help too much here. My understanding is that Avalonia does not currently have a Wayland "backend", and it looks like the Linux implementation of the Avalonia control (NativeControlHost) that I use to get a window handle to pass on to SDL2 seems to use e.g XCreateWindow to create its window.

Maybe this is helpful: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.X11/X11NativeControlHost.cs

**Comment by icculus** [comment]
Okay, so it sounds like we're good here for now. We'll revisit if this breaks when Avalonia starts doing Wayland things in the nebulous future.   :)
